base/utils/vmaas.go

```diff
@@ -0,0 +1,158 @@
+		if updateListA, ok := mergedList[nevraB]; ok {
+			merged, err := mergeUpdates(updateListA, updateListB)
+			if err != nil {
+				return nil, err
+			}
+			mergedList[nevraB] = *merged
+		} else {
+			mergedList[nevraB] = updateListB
+		}
+	}
+
+	vmaasDataA.UpdateList = &mergedList
+
+	return vmaasDataA, nil
+}
+
+func mergeUpdates(listA, listB vmaas.UpdatesV2ResponseUpdateList) (*vmaas.UpdatesV2ResponseUpdateList, error) {
+	updatesA, updatesB := listA.GetAvailableUpdates(), listB.GetAvailableUpdates()
+	if len(updatesA) == 0 && len(updatesB) == 0 {
+		return nil, nil
+	}
+	if len(updatesA) == 0 {
+		return &listB, nil
+	}
+	if len(updatesB) == 0 {
+		return &listA, nil
+	}
+
+	newUpdates, err := processMerge(updatesA, updatesB)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vmaas.UpdatesV2ResponseUpdateList{
+		AvailableUpdates: newUpdates,
+	}, nil
+}
+
+func processMerge(updatesA,
+	updatesB []vmaas.UpdatesV2ResponseAvailableUpdates) (*[]vmaas.UpdatesV2ResponseAvailableUpdates, error) {
+	newUpdates := make([]vmaas.UpdatesV2ResponseAvailableUpdates, 0)
+	processed := make(map[string]bool)
+	offset, i := 0, 0
+	for i < len(updatesA) {
+		nevraA, err := ParseNevra(updatesA[i].GetPackage())
+		if err != nil {
+			return nil, err
+		}
+		if offset > len(updatesB)-1 {
+			appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+		}
+		for j := 0 + offset; j < len(updatesB); j++ {
+			nevraB, err := ParseNevra(updatesB[j].GetPackage())
+			if err != nil {
+				return nil, err
+			}
+			cmp := nevraA.Cmp(nevraB)
+
+			if cmp < 0 {
+				appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+				break
+			}
+			if cmp > 0 {
+				appendUpdateIfNotExists(&newUpdates, updatesB[j], nevraB.Version, processed)
+				offset++
+				if offset == len(updatesB) {
+					appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+				}
+				continue
+			}
+			appendUpdateIfNotExists(&newUpdates, updatesA[i], nevraA.Version, processed)
+			appendUpdateIfNotExists(&newUpdates, updatesB[j], nevraB.Version, processed)
+			offset++
+			break
+		}
+		i++
+	}
+	if offset <= len(updatesB)-1 {
+		newUpdates = append(newUpdates, updatesB[offset:]...)
+	}
+	return &newUpdates, nil
+}
+
+func appendUpdateIfNotExists(list *[]vmaas.UpdatesV2ResponseAvailableUpdates,
+	update vmaas.UpdatesV2ResponseAvailableUpdates, version string, inserted map[string]bool) {
+	if !inserted[*update.Repository+version] {
+		*list = append(*list, update)
+		inserted[*update.Repository+version] = true
+	}
+}
+
+func ModuleListsEqual(listA *[]vmaas.UpdatesV3RequestModulesList, listB *[]vmaas.UpdatesV3RequestModulesList) bool {
+	if listA == nil || listB == nil {
+		return true
+	}
+
+	for _, moduleA := range *listA {
+		exists := false
+		for _, moduleB := range *listB {
+			if moduleA.ModuleName == moduleB.ModuleName {
+				exists = true
+			}
+		}
+		if !exists {
+			return false
+		}
+	}
+
```

Comment:
If the moduleB list is longer (a superset of moduleA) this function returns true. Which is not correct.